File: java-api-core/src/com/kinvey/java/sync/SyncManager.java
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */

package com.kinvey.java.sync;

import com.google.api.client.http.HttpRequest;
import com.google.api.client.json.GenericJson;
import com.google.gson.Gson;
import com.kinvey.java.AbstractClient;
import com.kinvey.java.Query;
import com.kinvey.java.cache.ICache;
import com.kinvey.java.cache.ICacheManager;
import com.kinvey.java.core.AbstractKinveyJsonClientRequest;
import com.kinvey.java.network.NetworkManager;
import com.kinvey.java.query.MongoQueryFilter;
import com.kinvey.java.store.BaseDataStore;
import com.kinvey.java.store.StoreType;
import com.kinvey.java.sync.dto.SyncCollections;
import com.kinvey.java.sync.dto.SyncItem;
import com.kinvey.java.sync.dto.SyncRequest;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Prots on 2/24/16.
 */
public class SyncManager {

    private static final String SYNC_ITEM_TABLE_NAME = "syncitems";

    private ICacheManager cacheManager;

    public SyncManager(ICacheManager cacheManager){
        this.cacheManager = cacheManager;
    }

    public void removeEntity(String collectionName, String curEntityID) {
        Query query = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        query.equals("collectionName", collectionName)
                .equals("meta._id", curEntityID);
        cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE).delete(query);
    }

    public List<String> getCollectionTables() {
        ICache<SyncCollections> collectionsCache  = cacheManager.getCache("syncCollections", SyncCollections.class, Long.MAX_VALUE);
        List<SyncCollections> collections = collectionsCache.get();
        List<String> ret = new ArrayList<String>();
        for (SyncCollections collection : collections){
            ret.add(collection.getCollectionName());
        }
        return ret;
    }

    /**
     * use {@link #popSingleItemQueue(String)}
     */
    @Deprecated
    public List<SyncRequest> popSingleQueue(String collectionName) {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder())
                .equals("collection", collectionName);
        List<SyncRequest> requests = requestCache.get(q);

        //delete request from the queue

        if (requests.size() > 0) {
            List<String> ids = new ArrayList<String>();
            for (SyncRequest request: requests){
                if(request != null){
                    ids.add(request.get("_id").toString());
                }
            }
            requestCache.delete(ids);
        }
        return requests;
    }

    public List<SyncItem> popSingleItemQueue(String collectionName) {
        ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder())
                .equals("collection", collectionName);
        return requestCache.get().size() !=0 ? requestCache.get(q) : null;
    }

    /**
     * use {@link #deleteCachedItem(String)}
     */
    @Deprecated
    public int deleteCachedRequest(String id) {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        return requestCache.delete(id);
    }

    public int deleteCachedItem(String id) {
        ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        return requestCache.delete(id);
    }

    public int deleteCachedItems(Query q) {
        ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        return requestCache.delete(q);
    }

    /**
     * use {@link #enqueueRequest(String, NetworkManager, RequestMethod, String)}
     */
    @Deprecated
    public void enqueueRequest(String collectionName, AbstractKinveyJsonClientRequest clientRequest) throws IOException {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        requestCache.save(createSyncRequest(collectionName, clientRequest));
    }

    public <T extends GenericJson> void enqueueRequest(String collectionName, NetworkManager<T> networkManager, RequestMethod method, String id) throws IOException {
        ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);

        Query entityQuery = AbstractClient.sharedInstance().query();
        entityQuery.equals("meta.id", id);
        List<SyncItem> itemsList = requestCache.get(entityQuery);

        if (itemsList.size() == 0) {
            requestCache.save(createSyncItem(collectionName, method, networkManager, id));
        }
    }

    /**
     * use {@link #enqueueRequests(String, NetworkManager, RequestMethod, List)}
     */
    @Deprecated
    public <T extends GenericJson> void enqueueRequests(String collectionName, NetworkManager<T> networkManager,  List<T> ret) throws IOException {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        List<SyncRequest> syncRequests = new ArrayList<>();
        for (T t : ret) {
            syncRequests.add(createSyncRequest(collectionName, networkManager.saveBlocking(t)));
        }
        requestCache.save(syncRequests);
    }

    public <T extends GenericJson> void enqueueRequests(String collectionName, NetworkManager<T> networkManager, RequestMethod method, List<T> ret) throws IOException {
        ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        List<SyncItem> syncRequests = new ArrayList<>();
        for (T t : ret) {
            Query entityQuery = AbstractClient.sharedInstance().query();
            String ID = (String) t.get("_id");
            entityQuery.equals("meta.id", ID);
            List<SyncItem> itemsList = requestCache.get(entityQuery);

            if (itemsList.size() == 0) {
                syncRequests.add(createSyncItem(collectionName, method, networkManager, (String) t.get("_id")));
            }
        }
        requestCache.save(syncRequests);
    }

    public <T extends GenericJson> void enqueueRequests(String collectionName, NetworkManager<T> networkManager, RequestMethod method, Iterable<String> ids) throws IOException {
        ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        List<SyncItem> syncRequests = new ArrayList<>();
        for (String id : ids) {
            syncRequests.add(createSyncItem(collectionName, method, networkManager, id));
        }
        requestCache.save(syncRequests);
    }

    private <T extends GenericJson> SyncItem createSyncItem(String collectionName, RequestMethod requestMethod, NetworkManager networkManager, String id) throws IOException {
        SyncRequest.SyncMetaData entityID = new SyncRequest.SyncMetaData();
        if (id != null) {
            entityID.id = id;
        }
        entityID.customerVersion = networkManager.getClientAppVersion();
        entityID.customheader = networkManager.getCustomRequestProperties() != null ?
                (String) networkManager.getCustomRequestProperties().get("X-Kinvey-Custom-Request-Properties") : null;

        return new SyncItem(
                requestMethod,
                entityID,
                collectionName
        );
    }

    public SyncRequest createSyncRequest(String collectionName, AbstractKinveyJsonClientRequest clientRequest) throws IOException {
        HttpRequest httpRequest = clientRequest.buildHttpRequest();
        SyncRequest.SyncMetaData entityID = new SyncRequest.SyncMetaData();

        if (httpRequest.getContent() != null) {
            ByteArrayOutputStream os = new ByteArrayOutputStream();
            httpRequest.getContent().writeTo(os);
            entityID.data = os.toString("UTF-8");
        }

        if (clientRequest.getJsonContent() != null) {
            entityID.id = clientRequest.getJsonContent().get("_id").toString();
        }
        if (clientRequest.containsKey("entityID")){
            entityID.id = clientRequest.get("entityID").toString();
        }
        entityID.customerVersion = clientRequest.getCustomerAppVersion();
        entityID.customheader = clientRequest.getCustomRequestProperties();

        return new SyncRequest(
                SyncRequest.HttpVerb.valueOf(clientRequest.getRequestMethod().toUpperCase()),
                entityID, httpRequest.getUrl(),
                collectionName
        );
    }

    public void enqueueRequest(SyncRequest request) {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        requestCache.save(request);
    }

    /**
     * This methods gets the count of sync operation to be performed
     * @param collectionName the name of the collection we want to get the info
     * @return the count of sync objects for given collection
     */
    public long getCount(String collectionName){
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        ICache<SyncItem> requestItemCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder())
                .equals("collection", collectionName);
        return requestCache.count(q) + requestItemCache.count(q);
    }

    /**
     * This method uses the AnsycAppData API to execute requests.  It contains an if/else block for the verb of the request
     * and then calls the appropriate appdata method.  As it uses the Async API, every request has a callback.
     * <p/>
     * Dependant on the verb and the result, this method will also update the local database.
     *
     * @param client kinvey client to execute request with
     * @param request Sync request to be executed
     */
    public void executeRequest(final AbstractClient client, SyncRequest request) throws IOException {

        client.setClientAppVersion(request.getEntityID().customerVersion);
        client.setCustomRequestProperties(new Gson().fromJson(request.getEntityID().customheader, GenericJson.class));

        GenericJson entity = null;
        try {
            if (request.getEntityID().data != null) {
                entity = client.getJsonFactory().createJsonParser(request.getEntityID().data).parse(GenericJson.class);

            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        BaseDataStore networkDataStore = BaseDataStore.collection(request.getCollectionName(), GenericJson.class, StoreType.NETWORK, client);

        if (request.getHttpVerb().equals(SyncRequest.HttpVerb.PUT) || request.getHttpVerb().equals((SyncRequest.HttpVerb.POST))) {

            if (entity != null){
                try {
                    networkDataStore.save(entity);
                } catch (Exception e){
                    enqueueRequest(request);
                    throw e;
                }
            }
        } else if (request.getHttpVerb().equals(SyncRequest.HttpVerb.DELETE)){
            String curID = request.getEntityID().id;

            if (curID != null && curID.startsWith("{") && curID.endsWith("}")){
                //it's a query
                Query q = new Query().setQueryString(curID);
                try {
                    networkDataStore.delete(q);
                } catch(Exception e) {
                    enqueueRequest(request);
                    throw e;
                }


            } else if (curID == null && request.getUrl().contains("?query=")) {

                String url = request.getUrl();
                int index = url.indexOf("?query=") + 7;
                String qString = url.substring(index);
                String decodedQuery = null;
                try {
                    decodedQuery = URLDecoder.decode(qString, "UTF-8");
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                }
                if (decodedQuery == null) {
                    return;
                }
                Query q = new Query().setQueryString(decodedQuery);
                try {
                    networkDataStore.delete(q);
                } catch(Exception e) {
                    enqueueRequest(request);
                    throw e;
                }

            } else {
                    //it's a single ID
                    try {
                        networkDataStore.delete(request.getEntityID().id);
                    } catch(Exception e) {
                        //TODO: need to check the errors
                        //enqueueRequest(request);
                        throw e;
                    }

                }
            }
    }

    public int clear(String collectionName) {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        ICache<SyncItem> requestItemCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder())
                .equals("collection", collectionName);
        return requestCache.delete(q) + requestItemCache.delete(q);
    }

    public int clear(String collectionName, Query query) {
        ICache<SyncRequest> requestCache = cacheManager.getCache("sync", SyncRequest.class, Long.MAX_VALUE);
        ICache<SyncItem> requestItemCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
        query = query.equals("collection", collectionName);
        return requestCache.delete(query) + requestItemCache.delete(query);
    }

    public ICacheManager getCacheManager() {
        return cacheManager;
    }
}

File: java-api-core/src/com/kinvey/java/sync/SyncManager.kt
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */

package com.kinvey.java.sync

import com.google.api.client.json.GenericJson
import com.google.api.client.json.JsonParser
import com.google.gson.Gson
import com.kinvey.java.AbstractClient
import com.kinvey.java.AbstractClient.Companion.sharedInstance
import com.kinvey.java.KinveySaveBatchException
import com.kinvey.java.Logger.Companion.ERROR
import com.kinvey.java.Query
import com.kinvey.java.cache.ICache
import com.kinvey.java.cache.ICacheManager
import com.kinvey.java.core.AbstractKinveyClientRequest
import com.kinvey.java.core.AbstractKinveyJsonClientRequest
import com.kinvey.java.model.KinveySyncSaveBatchResponse
import com.kinvey.java.network.NetworkManager
import com.kinvey.java.query.MongoQueryFilter.MongoQueryFilterBuilder
import com.kinvey.java.store.BaseDataStore
import com.kinvey.java.store.BaseDataStore.Companion.collection
import com.kinvey.java.store.StoreType
import com.kinvey.java.sync.dto.SyncCollections
import com.kinvey.java.sync.dto.SyncItem
import com.kinvey.java.sync.dto.SyncRequest
import com.kinvey.java.sync.dto.SyncRequest.HttpVerb
import com.kinvey.java.sync.dto.SyncRequest.SyncMetaData
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URLDecoder

/**
 * Created by Prots on 2/24/16.
 */
open class SyncManager(val cacheManager: ICacheManager?) {

    fun removeEntity(collectionName: String?, curEntityID: String?) {
        val query = Query(MongoQueryFilterBuilder())
        query.equals(COLLECTION_NAME, collectionName).equals(META_ID, curEntityID)
        cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)?.delete(query)
    }

    val collectionTables: List<String>
        get() {
            val collectionsCache = cacheManager?.getCache(SYNC_COLLECTIONS, SyncCollections::class.java, Long.MAX_VALUE)
            val collections = collectionsCache?.get()
            return collections?.mapNotNull { it.collectionName }?.toMutableList() as List<String>
        }

    @Deprecated("use [.popSingleItemQueue]")
    open fun popSingleQueue(collectionName: String?): List<SyncRequest>? {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        val q = Query(MongoQueryFilterBuilder()).equals(COLLECTION, collectionName)
        var requests: List<SyncRequest>? = null
        requestCache?.let {
            requests = it[q]
        }
        //delete request from the queue
        requests?.let { it ->
            val ids = it.map { request -> request[ID].toString() }.toMutableList()
            requestCache?.delete(ids)
        }
        return requests
    }

    fun popSingleItemQueue(collectionName: String?): List<SyncItem>? {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val q = Query(MongoQueryFilterBuilder()).equals(COLLECTION, collectionName)
        return requestCache?.run { if (this.get().isNotEmpty()) this[q] else null }
    }

    @Deprecated("use [.deleteCachedItem]")
    fun deleteCachedRequest(id: String?): Int {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        return requestCache?.delete(id ?: "") ?: 0
    }

    fun deleteCachedItem(id: String?): Int {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        return requestCache?.delete(id ?: "") ?: 0
    }

    fun deleteCachedItems(q: Query?): Int {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        return requestCache?.delete(q!!) ?: 0
    }

    @Deprecated("use [.enqueueRequest]")
    @Throws(IOException::class)
    fun enqueueRequest(collectionName: String, clientRequest: AbstractKinveyJsonClientRequest<*>) {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        requestCache?.save(createSyncRequest(collectionName, clientRequest))
    }

    @Deprecated("use [.enqueueRequests]")
    @Throws(IOException::class)
    fun <T : GenericJson> enqueueRequests(collectionName: String, networkManager: NetworkManager<T>, ret: List<T>) {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        val syncRequests = ret.map { item -> createSyncRequest(collectionName, networkManager.saveBlocking(item)) }
        requestCache?.save(syncRequests)
    }

    /**
     * Used to save sync requests that were not completed
     * @param request Sync request to be saved
     */
    fun enqueueRequest(request: SyncRequest) {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        requestCache?.save(request)
    }

    @Throws(IOException::class)
    fun <T : GenericJson> enqueueRequest(collectionName: String?, networkManager: NetworkManager<T>?, httpMethod: HttpVerb?, id: String?) {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val syncItem = prepareSyncItemRequest(requestCache, collectionName, networkManager, httpMethod, id)
        syncItem?.let { requestCache?.save(it) }
    }

    @Throws(IOException::class)
    fun <T : GenericJson> enqueueDeleteRequests(collectionName: String?, networkManager: NetworkManager<T>?, ret: List<T>?) {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val syncRequests = ret?.mapNotNull { item ->
            val syncItemId = item[ID] as String?
            prepareSyncItemRequest(requestCache, collectionName, networkManager, HttpVerb.DELETE, syncItemId)
        }
        syncRequests?.let { requests -> requestCache?.save(requests) }
    }

    @Throws(IOException::class)
    fun <T : GenericJson> enqueueSaveRequests(collectionName: String, networkManager: NetworkManager<T>, ret: List<T>?) {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val syncRequests = ret?.mapNotNull { item ->
            val syncItemId = item[ID] as String?
            prepareSyncItemRequest(requestCache, collectionName, networkManager, if (networkManager.isTempId(item)) HttpVerb.POST else HttpVerb.PUT, syncItemId)
        }
        syncRequests?.let { requests -> requestCache?.save(requests) }
    }

    @Throws(IOException::class)
    fun <T : GenericJson> enqueueDeleteRequests(collectionName: String?, networkManager: NetworkManager<T>?, ids: Iterable<String>?) {
        val requestCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val syncRequests = ids?.map { id -> prepareSyncItemRequest(requestCache, collectionName, networkManager, HttpVerb.DELETE, id) }
        syncRequests?.let { requests -> requestCache?.save(requests) }
    }

    @Throws(IOException::class)
    private fun <T : GenericJson> prepareSyncItemRequest(requestCache: ICache<SyncItem>?,
                                                          collectionName: String?,
                                                          networkManager: NetworkManager<T>?,
                                                          httpMethod: HttpVerb?,
                                                          syncItemId: String?): SyncItem? {
        val entityQuery = sharedInstance?.query()
        entityQuery?.equals(META_DOT_ID, syncItemId)
        val cache = requestCache ?: return null
        return entityQuery?.run {
            val itemsList = cache[this]
            when {
                itemsList.isEmpty() -> createSyncItem(collectionName, httpMethod, networkManager, syncItemId)
                httpMethod == HttpVerb.DELETE -> {
                    cache.delete(this)
                    createSyncItem(collectionName, httpMethod, networkManager, syncItemId)
                }
                else -> null
            }
        }
    }

    @Throws(IOException::class)
    private fun createSyncItem(collectionName: String?, httpMethod: HttpVerb?, networkManager: NetworkManager<*>?, id: String?): SyncItem {
        val entityID = SyncMetaData()
        if (id != null) {
            entityID.id = id
        }
        entityID.customerVersion = networkManager?.clientAppVersion
        entityID.customheader = networkManager?.customRequestProperties?.get("X-Kinvey-Custom-Request-Properties") as String?
        return SyncItem(httpMethod, entityID, collectionName)
    }

    @Throws(IOException::class)
    fun createSyncRequest(collectionName: String?, clientRequest: AbstractKinveyJsonClientRequest<*>?): SyncRequest {
        val httpRequest = clientRequest?.buildHttpRequest()
        val entityID = SyncMetaData()
        val os = ByteArrayOutputStream()
        httpRequest?.content?.writeTo(os)
        entityID.data = os.toString(UTF_8)
        if (clientRequest?.jsonContent != null) {
            entityID.id = clientRequest.jsonContent[ID].toString()
        }
        if (clientRequest?.containsKey(ENTITY_ID) == true) {
            entityID.id = clientRequest[ENTITY_ID].toString()
        }
        entityID.customerVersion = clientRequest?.customerAppVersion
        entityID.customheader = clientRequest?.customRequestProperties
        return SyncRequest(
                HttpVerb.valueOf(clientRequest?.requestMethod?.toUpperCase() ?: ""),
                entityID, httpRequest?.url, collectionName
        )
    }

    /**
     * This methods gets the count of sync operation to be performed
     * @param collectionName the name of the collection we want to get the info
     * @return the count of sync objects for given collection
     */
    fun getCount(collectionName: String?): Long {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        val requestItemCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val q = Query(MongoQueryFilterBuilder()).equals(COLLECTION, collectionName)
        return (requestCache?.count(q) ?: 0) + (requestItemCache?.count(q) ?: 0)
    }

    /**
     * This method uses the AnsycAppData API to execute requests.  It contains an if/else block for the verb of the request
     * and then calls the appropriate appdata method.  As it uses the Async API, every request has a callback.
     *
     *
     * Dependant on the verb and the result, this method will also update the local database.
     *
     * @param client kinvey client to execute request with
     * @param request Sync request to be executed
     */
    @Throws(IOException::class)
    open fun executeRequest(client: AbstractClient<*>?, request: SyncRequest?): GenericJson? {
        client?.clientAppVersion = request?.entityID?.customerVersion
        client?.setCustomRequestProperties(Gson().fromJson(request?.entityID?.customheader, GenericJson::class.java))
        var entity: GenericJson? = null
        var result: GenericJson? = null
        try {
            if (!request?.entityID?.data.isNullOrEmpty()) {
                entity = client?.jsonFactory?.createJsonParser(request?.entityID?.data)?.parse<GenericJson>(GenericJson::class.java)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        var networkDataStore: BaseDataStore<GenericJson>? = null
        if (client != null) {
            networkDataStore = collection(request?.collectionName ?: "", GenericJson::class.java, StoreType.NETWORK, client)
        }
        if (request?.httpVerb == HttpVerb.PUT || request?.httpVerb == HttpVerb.POST) {
            entity?.let {
                try {
                    if (request.httpVerb == HttpVerb.POST) {
                        // Remvove the _id, since this is a create operation
                        entity.set("_id", null)
                    }
                    result = networkDataStore?.save(it)
                } catch (e: Exception) {
                    enqueueRequest(request)
                    throw e
                }
            }
        } else if (request?.httpVerb == HttpVerb.DELETE) {
            val curID: String? = request.entityID?.id
            if (curID != null && curID.startsWith("{") && curID.endsWith("}")) {
                //it's a query
                val q = Query().setQueryString(curID)
                try {
                    networkDataStore?.delete(q)
                } catch (e: Exception) {
                    enqueueRequest(request)
                    throw e
                }
            } else if (curID == null && request.url?.contains("?query=") == true) {
                val url = request.url ?: ""
                val index = url.indexOf("?query=") + 7
                val qString = url.substring(index)
                var decodedQuery: String? = null
                try {
                    decodedQuery = URLDecoder.decode(qString, UTF_8)
                } catch (e: UnsupportedEncodingException) {
                    e.printStackTrace()
                }
                if (decodedQuery == null) {
                    return result
                }
                val q = Query().setQueryString(decodedQuery)
                try {
                    networkDataStore?.delete(q)
                } catch (e: Exception) {
                    enqueueRequest(request)
                    throw e
                }
            } else {
                //it's a single ID
                try {
                    networkDataStore?.delete(request.entityID?.id?: "")
                } catch (e: Exception) {
                    //TODO: need to check the errors
                    //enqueueRequest(request);
                    throw e
                }
            }
        }
        return result
    }

    /**
     * This method uses the AnsycAppData API to execute requests.  It contains an if/else block for the verb of the request
     * and then calls the appropriate appdata method.  As it uses the Async API, every request has a callback.
     *
     *
     * Dependant on the verb and the result, this method will also update the local database.
     *
     * @param client kinvey client to execute request with
     * @param request Sync request to be executed
     */
    @Throws(IOException::class)
    fun <T : GenericJson> executeBatchRequest(client: AbstractClient<*>, networkManager: NetworkManager<T>, request: SyncRequest): KinveySyncSaveBatchResponse<T>? {
        client.clientAppVersion = request.entityID?.customerVersion
        client.setCustomRequestProperties(Gson().fromJson(request.entityID?.customheader, GenericJson::class.java))
        var entityList: Collection<T>? = null
        val bunchData = request.entityID?.bunchData ?: false
        val dataJson: String? = request.entityID?.data
        try {
            if (dataJson != null) {
                val parser: JsonParser = client.jsonFactory.createJsonParser(dataJson)
                if (bunchData) {
                    entityList = parser.parseArray(List::class.java, networkManager.currentClass)
                } else return null
            }
        } catch (e: IOException) {
            ERROR(e.message)
        }
        val dataStoreNet: BaseDataStore<T> = collection(request.collectionName ?: "", networkManager.currentClass, StoreType.NETWORK, client)
        return runSaveBatchSyncRequest(request, entityList, dataStoreNet)
    }

    @Throws(IOException::class)
    private fun <T : GenericJson> runSaveBatchSyncRequest(request: SyncRequest,
                                                           entityList: Collection<T>?, dataStore: BaseDataStore<T>): KinveySyncSaveBatchResponse<T>? {
        var result: KinveySyncSaveBatchResponse<T>? = null
        if (entityList != null) {
            result = try {
                val resultItems: List<T> = dataStore.saveBatch(entityList)
                KinveySyncSaveBatchResponse(resultItems, null)
            } catch (e: KinveySaveBatchException) {
                KinveySyncSaveBatchResponse<T>(e.entities as List<T>, e.errors)//throw e;
            } catch (e: Exception) {
                enqueueRequest(request)
                throw e
            }
        }
        return result
    }

    @Throws(IOException::class)
    fun <T : GenericJson> createSaveBatchSyncRequest(collectionName: String?, networkManager: NetworkManager<T>, ret: List<T>?): SyncRequest {
        val request = createSaveBatchSyncRequest(collectionName, networkManager.saveBatchBlocking(ret) as AbstractKinveyClientRequest<*>)
        request.entityID?.bunchData = true
        return request
    }

    @Throws(IOException::class)
    fun createSaveBatchSyncRequest(collectionName: String?, clientRequest: AbstractKinveyClientRequest<*>): SyncRequest {
        val httpRequest = clientRequest.buildHttpRequest()
        val entityID = SyncMetaData()
        if (httpRequest.content != null) {
            val os = ByteArrayOutputStream()
            httpRequest.content.writeTo(os)
            entityID.data = os.toString(UTF_8)
        }
        if (clientRequest is AbstractKinveyJsonClientRequest<*>) {
            val content = clientRequest.jsonContent
            if (content != null) {
                entityID.id = content[ID].toString()
            }
        }
        if (clientRequest.containsKey(ENTITY_ID)) {
            entityID.id = clientRequest[ENTITY_ID].toString()
        }
        entityID.customerVersion = clientRequest.customerAppVersion
        entityID.customheader = clientRequest.customRequestProperties
        val requestMethod = clientRequest.requestMethod.toUpperCase()
        return SyncRequest(
                HttpVerb.fromString(requestMethod),
                entityID, httpRequest.url,
                collectionName
        )
    }

    fun clear(collectionName: String?): Int {
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        val requestItemCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        val q = Query(MongoQueryFilterBuilder()).equals(COLLECTION, collectionName)
        return (requestCache?.delete(q) ?: 0) + (requestItemCache?.delete(q) ?: 0)
    }

    fun clear(collectionName: String?, query: Query): Int {
        var query = query
        val requestCache = cacheManager?.getCache(SYNC, SyncRequest::class.java, Long.MAX_VALUE)
        val requestItemCache = cacheManager?.getCache(SYNC_ITEM_TABLE_NAME, SyncItem::class.java, Long.MAX_VALUE)
        query = query.equals(COLLECTION, collectionName)
        return (requestCache?.delete(query) ?: 0) + (requestItemCache?.delete(query) ?: 0)
    }

    companion object {
        private const val SYNC_ITEM_TABLE_NAME = "syncitems"
        private const val COLLECTION = "collection"
        private const val COLLECTION_NAME = "collectionName"
        private const val SYNC = "sync"
        private const val META_DOT_ID = "meta.id"
        private const val META_ID = "meta._id"
        private const val ID = "_id"
        private const val UTF_8 = "UTF-8"
        private const val ENTITY_ID = "entityID"
        private const val SYNC_COLLECTIONS = "syncCollections"
    }

}